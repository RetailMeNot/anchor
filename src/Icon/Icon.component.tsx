// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled from 'styled-components';
// COMPONENTS
import {
  Hamburger,
  ChevronDown,
  ChevronUp,
  ChevronRight,
  Cart,
  Search
} from './SVGs';
// THEME

type IconTypes = 'hamburger' | 'chevron-down' | 'chevron-up' | 'chevron-right' | 'cart' | 'search';

export interface IconSVGProps {
  width?: number;
  height?: number;
  color?: string;
}

interface IconProps {
  type: IconTypes;
  children?: any;
  className?: string;
  color?: string;
  width?: number;
  height?: number;
}

export const StyledIcon = styled.div`
  display: inline-flex;
  max-width: 1.5rem;
  justify-content: end;

  > svg {
    flex: 0 0 1.25rem;
  }
`;

const DefaultProps: IconProps = {
  type: 'hamburger', // TODO: default icon
  color: '#222',
};

const Icons = {
  'hamburger': Hamburger,
  'chevron-down': ChevronDown,
  'chevron-up': ChevronUp,
  'chevron-right': ChevronRight,
  'cart': Cart,
  'search': Search,
};

export const Icon = ({
  className,
  type,
  color,
  width,
  height,
  ...props
}: IconProps = DefaultProps): React.ReactElement<any> => (
  <StyledIcon className={classNames(className)} {...props}>
    {Icons[type]({ color, width, height })}
  </StyledIcon>
);

Icon.defaultProps = DefaultProps;

export default Icon;
