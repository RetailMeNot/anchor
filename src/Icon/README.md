# Icon

The Icon component is actually an object which holds all Icon components. 
Trying to render `<Icon />` will result in an error, but for example, `<Icon.ArrowBack />` would not.

## Usage
    import { Icon } from '@retailmenot/anchor';
    
    const example = props => (
        <Icon.ArrowBack />
    );
    
Alternatively, each icon can be imported individually.
    
    import { IconArrowBack } from '@retailmenot/anchor';
    
    const example = props => (
        <IconArrowBack />
    );    
    
## Settings/Props

The component can accept 3 `props`:

* `color`: Defaults to **`#222`** but if passed it will take whatever hex/rgb/rgba value provided.
* `scale`: Defaults to **`md`**. The full range of options are **`'xs', 'sm', 'md', 'lg', 'xl', 'xxl'`**
* `className`: An additional class name. All icons have the **`anchor-icon`** class name in addition to a more specific class.
Ex: **`class="anchor-icon arrow-back"`**.    
        
    

