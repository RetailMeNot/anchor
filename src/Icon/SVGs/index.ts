export { default as Hamburger } from './Hamburger.svg';

export { default as ChevronDown } from './ChevronDown.svg';
export { default as ChevronUp } from './ChevronUp.svg';
export { default as ChevronRight } from './ChevronRight.svg';

export { default as Cart } from './Cart';
export { default as Search } from './Search';
