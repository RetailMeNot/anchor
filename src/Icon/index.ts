export { default as Icon } from './Icon.component';
