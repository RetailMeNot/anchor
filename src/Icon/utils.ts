// VENDOR
import styled from '@xstyled/styled-components';
import { margin as marginStyles } from '../theme/system.theme';

type ScaleFactors = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl';

export interface IconSVGProps {
    scale?: ScaleFactors;
    color?: string;
    className?: string;
    children?: any;
}

export const Scale = {
    xs: 8,
    sm: 14,
    md: 16,
    lg: 24,
    xl: 32,
    xxl: 48,
};

export const DefaultColor = 'currentColor';
export const DefaultScale = 'md';

export const StyledIcon = styled.span<IconSVGProps>`
    ${marginStyles}
    display: inline-block;
    height: ${({ scale = 'md' }: IconSVGProps) => `${Scale[scale] / 16}rem`};
    width: ${({ scale = 'md' }: IconSVGProps) => `${Scale[scale] / 16}rem`};
    line-height: 0;
`;

StyledIcon.displayName = 'AnchorIcon';

export type AnchorIcons =
    | 'AddEvent'
    | 'ArrowBack'
    | 'ArrowForward'
    | 'AvatarIcon'
    | 'AvatarOutline'
    | 'BarCode'
    | 'BulletList'
    | 'Calendar'
    | 'Camera'
    | 'Cart'
    | 'CashBack'
    | 'Cells'
    | 'Chat'
    | 'Check'
    | 'CheckSmall'
    | 'ChevronDown'
    | 'ChevronDownSmall'
    | 'ChevronLeft'
    | 'ChevronLeftSmall'
    | 'ChevronRight'
    | 'ChevronRightSmall'
    | 'ChevronUp'
    | 'ChevronUpSmall'
    | 'Clock'
    | 'Close'
    | 'CloseSmall'
    | 'CommentMore'
    | 'CreditCard'
    | 'CrossHairs'
    | 'Cut'
    | 'Disabled'
    | 'Dislike'
    | 'Download'
    | 'Ellipses'
    | 'EllipsesVertical'
    | 'Envelope'
    | 'EnvelopeOpen'
    | 'Error'
    | 'Expand'
    | 'Gear'
    | 'GiftCard'
    | 'Hamburger'
    | 'Heart'
    | 'HeartOutline'
    | 'Home'
    | 'Info'
    | 'InfoOutline'
    | 'Laptop'
    | 'Lightning'
    | 'Like'
    | 'ListIcon'
    | 'Lock'
    | 'Map'
    | 'Marker'
    | 'MarkerOutline'
    | 'Mobile'
    | 'News'
    | 'Pencil'
    | 'Play'
    | 'Plus'
    | 'PlusSmall'
    | 'Print'
    | 'Question'
    | 'QuestionOutline'
    | 'Refresh'
    | 'RetailMeNotLogo'
    | 'SadFace'
    | 'Search'
    | 'Share'
    | 'Sliders'
    | 'Star'
    | 'StarHalf'
    | 'StarOutline'
    | 'Success'
    | 'SuccessOutline'
    | 'Tag'
    | 'TagAdd'
    | 'Upload';
