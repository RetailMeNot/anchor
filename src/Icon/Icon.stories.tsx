// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled from 'styled-components';
// COMPONENT
import Icon from './Icon.component';

const StyledStory = styled.div`
  padding: 2rem 5rem;
`;

storiesOf('Components/Icon', module)
  .add('All', () => (
      <StyledStory>
        <Icon type="hamburger" />
      </StyledStory>
  ));
