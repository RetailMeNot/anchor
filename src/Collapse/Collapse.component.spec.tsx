// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// Vendor
import { shallow, mount } from 'enzyme';
// COMPONENT
import { Collapse } from './';
import { ArrowBack } from '../Icon';

describe('Component: Collapse', () => {
    it('should match its snapshot.', () => {
        const subject = <Collapse />;
        const component = shallow(subject);

        expect(subject).toBeDefined();
        expect(component).toBeDefined();

        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should show content when isOpen is true.', () => {
        const subject = <Collapse isOpen={true}>Hello World!</Collapse>;
        const component = shallow(subject);

        expect(
            component.find('.anchor-collapse-content').exists()
        ).toBeTruthy();
    });

    it('should not show content by default.', () => {
        const subject = <Collapse>Hello World!</Collapse>;
        const component = shallow(subject);

        expect(component.find('.anchor-collapse-content').exists()).toBeFalsy();
    });

    it('should use custom openedText as closedText if no closedText is given.', () => {
        const sampleText = 'No Toggling Text';
        const subject = (
            <Collapse openedText={sampleText} isOpen={true}>
                Hello World!
            </Collapse>
        );
        const component = shallow(subject);

        expect(
            component.find('.anchor-collapse-button').contains(sampleText)
        ).toBeTruthy();
    });

    it('should toggle content when the button is clicked.', () => {
        const subject = <Collapse>Hello World!</Collapse>;
        const wrapper = mount(subject);
        const toggleButton = wrapper.find('.anchor-collapse-button');

        toggleButton.simulate('click');

        expect(wrapper.find('.anchor-collapse-content').exists()).toBeTruthy();
    });

    it('should use different closed icon when passed as a prop.', () => {
        const subject = (
            <Collapse closedIcon={<ArrowBack />}>Hello World!</Collapse>
        );
        const wrapper = mount(subject);

        expect(wrapper.find('.anchor-icon.arrow-back').exists()).toBeTruthy();
    });

    it('should use different opened icon when passed as a prop.', () => {
        const subject = (
            <Collapse openedIcon={<ArrowBack />} isOpen={true}>
                Hello World!
            </Collapse>
        );
        const wrapper = mount(subject);

        expect(wrapper.find('.anchor-icon.arrow-back').exists()).toBeTruthy();
    });
});
