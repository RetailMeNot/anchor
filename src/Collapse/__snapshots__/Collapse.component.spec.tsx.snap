// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Collapse should match its snapshot. 1`] = `
<div
  className="anchor-collapse comfortable sc-bwzfXH bcZeDR"
>
  <button
    className="anchor-collapse-button"
    onClick={[Function]}
  >
    Open
     
    <span
      className="anchor-icon chevron-up sc-bdVaJa gENAGr"
      scale="md"
    >
      <svg
        height={16}
        viewBox="0 0 16 16"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 7.523l2.954 2.518a.77.77 0 0 0 1.058-.001l.014-.014a.721.721 0 0 0-.009-1.052L8.53 5.96a.767.767 0 0 0-1.048-.009l-3.506 3.03a.719.719 0 0 0-.001 1.043l.015.015a.767.767 0 0 0 1.048.01L8 7.523z"
          fill="currentColor"
          fillRule="nonzero"
        />
      </svg>
    </span>
  </button>
</div>
`;
